File: matlab/classifiers/classifierGMM_EM.m
function [Ytp,classifier] = classifierGMM_EM(Zs,ks,Zt,classifier)
% (Unsupervised) Expectation maximisation (EM) Gaussian mixture model
% (maximum likelihood estimates)
%
% Inputs 
% Zs = source data
% ks = no. of source components
% Zt = target data
% classifier = pretrained classifier
%
% Outputs
% Ytp = target label predictions
% classifier = trained classifier

if nargin <4
    % Train unsupervised GMM Classifier
    classifier = gmm_mle_em(Zs,ks);
end

% Predict
Ytp = gmm_mle_predict(classifier,Zt);

end

File: classifiers/classifierGMM_EM.py
from sklearn.mixture import GaussianMixture
import numpy as np

def classifierGMM_EM(Zs, ks, Zt, classifier=None):
    """
    (Unsupervised) EM Gaussian mixture model

    Inputs:
    Zs: source data
    ks: number of components
    Zt: target data
    classifier: pretrained classifier

    Outputs:
    Ytp: target predictions
    classifier: trained classifier
    """
    if classifier is None:
        classifier = GaussianMixture(n_components=ks, random_state=0)
        classifier.fit(Zs)
    Ytp = classifier.predict(Zt)
    return Ytp, classifier